### ISSUE TITLE
Access to galaxy `position_angle` using standard sign convention

### ISSUE THREAD

**Issue by jchiang87** [LSSTDESC/gcr-catalogs#607]
For DC2, the imSim code assumed that the position angle (PA) for galaxies in cosmoDC2 follows the PhoSim convention of defining the PA as positive to the West of North.  This is similar to the sign convention that PhoSim uses for the `shear2` quantity, which is the opposite of the treecorr convention.  So, in order to use the standard convention for galaxy position angles for which the PA increases to the East, we are requesting an entry in `CosmoDC2GalaxyCatalog._generate_quantity_modifiers` that flips the sign for `position_angle` and `position_angle_true`, and adds quantities `position_angle_phosim` and `position_angle_true_phosim`, so that the PhoSim versions are explicitly available.

**Comment by JoanneBogart** [comment]
@yymao are you ok with this?   If so I can look into implementing it.

**Comment by yymao** [comment]
Yes that sounds good.

**Comment by evevkovacs** [comment]
Or I can do it too. Seems pretty straightforward.

**Comment by JoanneBogart** [comment]
@evevkovacs that would be great. Thanks!

**Comment by evevkovacs** [comment]
@jchiang87  I started to implement this but I am a bit confused. There is a quantity called 'position_angle_true', but I don't see one called position_angle.  There is a native quantity called 'morphology/positionAngle' (which is the angle of the axis of the ellipse on the sky), which is used for position_angle_true.
I think the correct thing to do here is to just define 'position_angle_true_phosim', as (np.negative, 'position_angle_true').
Is that ok?

**Comment by jchiang87** [comment]
In the [schema](https://github.com/LSSTDESC/gcr-catalogs/blob/master/GCRCatalogs/SCHEMA.md), I see both `position_angle` and `position_angle_true`.   How does the `position_angle` listed there get set?   Maybe @JoanneBogart could comment since it's the skyCatalogs code that would read that value using gcr-catalogs?

**Comment by evevkovacs** [comment]
Ah, so it may be in the schema, but we don't supply it in cosmoDC2. I assume it is supposed to be the lensed value of position_angle_true, and apparently we are not computing that.

**Comment by yymao** [comment]
`position_angle` and `position_angle_true` are for lensed and unlensed values, respectively.

**Comment by evevkovacs** [comment]
I can verify with @patricialarsen

**Comment by jchiang87** [comment]
@evevkovacs It looks like skyCatalogs just uses `position_angle_true`, so doing what you proposed should be what we need.

**Comment by evevkovacs** [comment]
OK, great thanks. I'll just verify that the new code works and then commit it.

**Comment by jchiang87** [comment]
Actually, I think we need something like
```
'position_angle_true_phosim':      (_gen_position_angle, 'morphology/positionAngle'),
'position_angle_true':             (np.negative, 'position_angle_true_phosim'),
```
to be consistent with what we did with DC2, i.e., the current PA values are the phosim versions.  (Not sure if the above code would actually work, but that's the idea.)

**Comment by JoanneBogart** [comment]
Sorry I didn't see this earlier, but just to confirm: skyCatalogs needs only `position_angle_true`.

**Comment by yymao** [comment]
Just out of curiosity, how does skycatalog incorporate lensing effect?

**Comment by jchiang87** [comment]
Lensing is applied in imSim using galsim to apply the WL parameters.

**Comment by evevkovacs** [comment]
@jchiang87  If I change the definition as you suggested above, that will flip the sign of PA in cosmoDC2 going forward. Is that what you really want?  Won't that be confusing?

**Comment by jchiang87** [comment]
I think this is the same situation as `shear_2`, where it has also been flipped.  The issue is that imSim v1.0 assumed that the phosim convention for position angles needed a sign flip in order to obtain an angle with the standard convention. In imSim v2.0, we want to get the position angle without needing to apply that extra sign change, so that means it has to move upstream.  Applying it here means that no one else needs to apply the flip.

**Comment by yymao** [comment]
I am ok with this change. But as Eve pointed out, this will result in a backward imcompabibltiy for anyone who is using `position_angle_true` currently, so we want to make sure that we advertise it to our users (including public users as this will affect them as well).

**Comment by evevkovacs** [comment]
OK, thank you, just double checking.

**Comment by jchiang87** [comment]
Thank you both for your help.

**Comment by evevkovacs** [comment]
@yymao In the reader, cosmodc2. py,  do you recall why the function _gen_position_angle is used? I don't think that I wrote this function and was wondering where it came from. Thanks

**Comment by yymao** [comment]
`_gen_position_angle` was used in early version of comosDC2 where position angles were not available in the catalog, and were generated randomly on the fly.

**Comment by evevkovacs** [comment]
So, in this case, I propose to eliminate this function and simply use the catalog value.

**Comment by yymao** [comment]
But then the reader will not work with those older versions of cosmoDC2 will no longer work. I am not sure if we want to make legacy catalogs not working even though they have been deprecated?

**Comment by evevkovacs** [comment]
I think that the function is not working as intended or at least, behave the way I think it should: ie.  if values of position angle are supplied in the catalog, then it should use those. But that's not what happens. When I tested it, the function did not return the values of 'morphology/positionAngle'.  If we agree that the desired behavior is what I just described, then I will look into revising the function so that it works the way I just outlined. What do you all think?

**Comment by yymao** [comment]
I agree that's behavior we want. 

I just looked at the code again -- it seems that maybe we only need this function for AlphaQ but we accidentally copied it to cosmoDC2? If that's the case then indeed we should remove it for cosmoDC2.

**Comment by evevkovacs** [comment]
That is certainly possible. AFAIK, we have always supplied 'morphology/positionAngle' for cosmoDC2.

**Comment by JoanneBogart** [comment]
I've always wondered why a random number was generated when there seemed to be a perfectly good value in the cosmoDC2 file natively!  
Does `morphology/positionAngle` follow the phoSim convention?  Either way, we still might need two values, one that is explicitly identified as following the phoSim convention, say `position_angle_true_phosim`, and then "regular" `position_angle_true`
Also - what will you do about `ellipticity_1_true` and `ellipticity_2_true`?  They're currently calculated using _gen_pos_angle.  Will they be changed to use `morphology/positionAngle`?   In skyCatalogs we're planning to store these values, so we need an `ellipticity_2_true` which does *not* follow the phoSim convention.  (@jchiang87 please comment/correct if need be.)

**Comment by jchiang87** [comment]
I would suggest the following:
* Update the code to provide the intended values for position angle and related quantities, which sounds like that means using the existing catalog values.
* Also provide columns that give the DC2 values, e.g., `position_angle_true_dc2`.  This avoids having a completely separate legacy catalog.
Given that the values we used for DC2 aren't even related to the intended values, identifying them as the "phosim" versions seems irrelevant.  skyCatalogs should just use the intended values, which, I hope will use the standard convention for the sign, i.e., not the phosim convention.

**Comment by yymao** [comment]
Are the position angles that are stored in the catalog files also just uniformly at random?  If so, wouldn't it be better to keep the current random position angles for legacy/consistency reason rather than using a new set of randoms?

**Comment by evevkovacs** [comment]
Yes, the position angles in the catalog are also uniformly random. However, the gen_position_angle function has always been superceding the catalog values, so for legacy/consistency, I propose that we do the following:

'position_angle_true_phosim':       'morphology/positionAngle',
'position_angle_true':             (np.negative, 'morphology/positionAngle'),
'position_angle_true_dc2':  (_gen_position_angle, 'morphology/positionAngle'),  #legacy code for dc2

Also, as Joanne pointed out, to take care of ellipticity_1_true and ellipticity_2_true, I propose to define
ellipticity_1_true_dc2 and ellipticity_2_true_dc2 which will supply the legacy values and define
ellipticity_1_true and ellipticity_2_true based on position_angle_true (treecorr convention)

Going forward, I think this gives us the behavior we want, but also allows the recovery of the DC2 values, as Jim suggested.

Does that sound reasonable to everyone?

**Comment by yymao** [comment]
It sounds fine but is there any reason we need to revert to the catalog values (since it's also uniformly random)?

**Comment by JoanneBogart** [comment]
Maybe it's just an aesthetic preference, but I would say GCRCatalogs should use values from the underlying catalog whenever they're available and suitable. It makes it easier to understand what's going on and there is that much less code to maintain.

**Comment by yymao** [comment]
I agree with this general principle. But given that the current values have been used in DC2, I feel it can actually create more confusion. 

Ideally we can create a new set of cosmoDC2 files with the existing values written into the files, and then update the reader to remove the random generation. But that might be too much work.

Anyway, I am ok with proceeding with the plan Eve outlined above, and it seems Jim and Joanne have already agreed too. I just want to register my warning here because I think these `_dc2` columns can potentially be a headache in the future.

**Comment by JoanneBogart** [comment]
People wishing to duplicate previous results or make comparisons to them should be warned about the different behavior of different GCRCatalogs versions. We could provide instructions for using an older version.

**Comment by jchiang87** [comment]
Closed by #608
